tiatoolbox/models/segmentation/semantic_segmentor.py

```diff
@@ -0,0 +1,983 @@
+
+    @staticmethod
+    def collate_fn(batch):
+        """Prototype to handle reading exception.
+
+        This will exclude any sample with `None` out from the
+        batch. As such, wrapping `__getitem__` with try-catch
+        and return `None` upon exceptions will prevent crashing the
+        entire programs. But as side effect, batch may not have size
+        as defined.
+
+        """
+        batch = [v for v in batch if v is not None]
+        return torch.utils.data.dataloader.default_collate(batch)
+
+    def __getitem__(self, idx: int):
+        # ! no need to lock as we dont modify source value in shared space
+        if self.wsi_idx != self.mp_shared_space.wsi_idx:
+            self.wsi_idx = int(self.mp_shared_space.wsi_idx.item())
+            self.reader = self._get_reader(self.wsi_paths[self.wsi_idx])
+
+        # this is in XY and at requested resolution not baseline
+        bound = self.mp_shared_space.patch_inputs[idx]
+        bound = bound.numpy()  # expected to be torch.Tensor
+
+        # be the same as bounds br-tl, unless bounds are of float
+        patch_data_ = []
+        scale_factors = self.ioconfig.scale_to_highest(
+            self.ioconfig.input_resolutions, self.ioconfig.resolution_unit
+        )
+        for idy, resolution in enumerate(self.ioconfig.input_resolutions):
+            resolution_bound = np.round(bound * scale_factors[idy])
+            patch_data = self.reader.read_bounds(
+                resolution_bound.astype(np.int32),
+                coord_space="resolution",
+                pad_constant_values=0,  # expose this ?
+                **resolution,
+            )
+
+            if self.preproc is not None:
+                patch_data = patch_data.copy()
+                patch_data = self.preproc(patch_data)
+            patch_data_.append(patch_data)
+        if len(patch_data_) == 1:
+            patch_data_ = patch_data_[0]
+
+        bound = self.mp_shared_space.patch_outputs[idx]
+        return patch_data_, bound
+
+
+class SemanticSegmentor:
+    """Pixel-wise segmentation predictor.
+
+    Note, if model is supplied in the arguments, it will override the backbone.
+
+    Args:
+        model (nn.Module): Use externally defined PyTorch model for prediction with.
+            weights already loaded. Default is `None`. If provided,
+            `pretrained_model` argument is ignored.
+        pretrained_model (str): Name of the existing models support by tiatoolbox
+            for processing the data. Refer to
+            `tiatoolbox.models.classification.get_pretrained_model` for details.
+            By default, the corresponding pretrained weights will also be
+            downloaded. However, you can override with your own set of weights
+            via the `pretrained_weight` argument. Argument is case insensitive.
+        pretrained_weight (str): Path to the weight of the corresponding
+            `pretrained_model`.
```

Comment:
Should this be singular? Is there only one weight or a bunch of them in the file?

Reply:
I dont understand what are you saying here. Each *model* in runtime effectively has only 1 set of weight. If there is another one then its already a different model. To be more clear
- Architecture definition: `pretrained_model`
- Set of weight to loaded: `pretrained_weigth`
These you combined into *model*

Reply:
So if it is a set of weights it should be plural 'weights' not 'weight'